dev-amdgpu: Update GPU SDMA model to support non-zero VMID packets, MMIOs, and fix SDMA poll livelock

v-ramadas opened the pull request:
This PR adds three fixes:

1. Fix a livelock in GPU SDMA where the SDMA polls memory locations continuously without allowing the Linux drivers to update the location
2. Add support for handling MMIO writes to non-zero VMID page tables
3. Add support for translation non-zero VMID SDMA packets

v-ramadas committed:
dev-amdgpu: Add support for non-zero VMID page table MMIOs

This commit adds support for non-zero VMID page table MMIOs that are
crucial to find the correct translations for SDMA packets with non-zero
VMIDs

v-ramadas commented:
@abmerop, I've included a commit to add delays to SDMA read/write operations in this PR.

abmerop commented on src/dev/amdgpu/amdgpu_nbio.hh:
```diff
@@ -103,6 +103,20 @@ class AMDGPUDevice;
 #define MI300X_INV_ENG17_ACK11                            0x1e2f98
 #define MI300X_EPF0_STRAP0                                0x34d8
 
+#define IS_regBM_PAGE_TABLE_BASE_ADDR(offset) \
+        ((offset >= 0x6b0ac && offset <=0x6b128) ? true : false)
+#define GET_CONTEXT_FROM_regBM_PAGE_TABLE_BASE_ADDR(offset) \
+        (((offset - 0x6b0ac) >> 2)/2)
+#define IS_regBM_PAGE_TABLE_START_ADDR(offset) \
+        ((offset >= 0x6b12c && offset <=0x6b1a8) ? true : false)
+#define GET_CONTEXT_FROM_regBM_PAGE_TABLE_START_ADDR(offset) \
+        (((offset - 0x6b12c) >> 2)/2)
+#define IS_regBM_PAGE_TABLE_END_ADDR(offset) \
+        ((offset >= 0x6b1ac && offset <=0x6b1c8) ? true : false)
+#define GET_CONTEXT_FROM_regBM_PAGE_TABLE_END_ADDR(offset) \
+        (((offset - 0x6b1ac) >> 2)/2)
```
These need to have MI200 somewhere in the name. The offsets here are different in MI300.

It might be better to have these be class methods. These can't be resolved at compile time anyway and the method would more easily allow for checking the gfx version and selecting the appropriate MMIO addresses.

v-ramadas replied:
Done

abmerop commented on src/dev/amdgpu/sdma_engine.cc:
```diff
@@ -366,7 +366,7 @@ SDMAEngine::decodeNext(SDMAQueue *q)
             [ = ] (const uint32_t &header)
                 { decodeHeader(q, header); });
-        dmaReadVirt(q->rptr(), sizeof(uint32_t), cb, &cb->dmaBuffer);
+        dmaReadVirt(q->rptr(), sizeof(uint32_t), cb, &cb->dmaBuffer, 1e9);
```
`1e9` should probably be defined as a `static constexpr int` in some header file so we could play with the value more easily if it ends up being a latency bottleneck in the future. Same applies to the usages elsewhere.

v-ramadas replied:
Done

mattsinc commented on src/dev/amdgpu/amdgpu_nbio.cc:
```diff
@@ -204,14 +204,51 @@ AMDGPUNbio::writeMMIO(PacketPtr pkt, Addr offset)
+        uint16_t context_id =
+            GET_CONTEXT_FROM_regBM_PAGE_TABLE_BASE_ADDR(offset);
+        regs[offset] = pkt->getLE<uint32_t>();
+        if ((offset % 8) == 0) {
```
add comment explaining what 8 means

v-ramadas replied:
Done

mattsinc commented on src/dev/amdgpu/sdma_engine.cc:
```diff
@@ -461,7 +464,7 @@ SDMAEngine::decodeHeader(SDMAQueue *q, uint32_t header)
                 [ = ] (const uint64_t &)
                     { write(q, (sdmaWrite *)dmaBuffer); });
-            dmaReadVirt(q->rptr(), sizeof(sdmaWrite), cb, dmaBuffer);
+            dmaReadVirt(q->rptr(), sizeof(sdmaWrite), cb, dmaBuffer, 1e9);
```
not sure if it's a GitHub oddity, but spacing looks wrong here (and on line 431 above)

v-ramadas replied:
I don't see this in my editor. Must be a GitHub oddity? I remember this happening earlier too

v-ramadas committed:
dev-amdgpu: Add a small delay to SDMA operations

The SDMA operations are atomic and cause livelocks in some cases,
especially when polling memory regions, when they are scheduled with
zero delay. This commit adds a small delay to fix livelocks

v-ramadas committed:
dev-amdgpu: Handle SDMA packets with non-zero VMIDs

Packets with VMID == 0 use GART to translate addresses. Packets with
VMID > 0 use page tables for translations. This commit updates the SDMA
model so that it does not perform GART translations for non-zero VMIDs.
This commit also sets the VMID value from packet headers wherever
necessary.

abmerop reviewed (commented):
I think it would be a good idea still to either `#define` the MMIO addresses or `static constexpr int` them.

I know this is leading towards greater work, but this would fix a potential future style issue

mattsinc commented:
> I think it would be a good idea still to either `#define` the MMIO addresses or `static constexpr int` them.
> 
> I know this is leading towards greater work, but this would fix a potential future style issue

we should just fix it now before merging in then

v-ramadas commented:
Thanks for the suggestions! I've moved the register addresses to `#define` macros.

mattsinc reviewed (approved):
This appears to address @abmerop 's comments, but will let him chime in before merging
